OpenRA.Mods.Common/Traits/Render/WithMakeAnimation.cs

```diff
@@ -18,39 +18,70 @@
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Replaces the sprite during construction.")]
-	public class WithMakeAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>
+	public class WithMakeAnimationInfo : ITraitInfo, Requires<WithSpriteBodyInfo>, Requires<UpgradeManagerInfo>
 	{
-		[Desc("Sequence name to use")]
+		[Desc("Sequence name to use.")]
 		[SequenceReference] public readonly string Sequence = "make";
 
+		[UpgradeGrantedReference]
+		[Desc("The upgrades to grant while the make animation runs.")]
+		public readonly string[] MakeUpgrades = { };
+
 		public object Create(ActorInitializer init) { return new WithMakeAnimation(init, this); }
 	}
 
-	public class WithMakeAnimation
+	public class WithMakeAnimation : INotifyCreated
 	{
 		readonly WithMakeAnimationInfo info;
 		readonly WithSpriteBody wsb;
+		readonly UpgradeManager manager;
 
 		public WithMakeAnimation(ActorInitializer init, WithMakeAnimationInfo info)
 		{
 			this.info = info;
 			var self = init.Self;
 			wsb = self.Trait<WithSpriteBody>();
+			manager = self.Trait<UpgradeManager>();
+		}
 
+		public void Created(Actor self)
+		{
 			var building = self.TraitOrDefault<Building>();
 			if (building != null && !building.SkipMakeAnimation)
 			{
+				foreach (var up in info.MakeUpgrades)
+					manager.GrantUpgrade(self, up, this);
+
 				wsb.PlayCustomAnimation(self, info.Sequence, () =>
 				{
 					building.NotifyBuildingComplete(self);
+					foreach (var up in info.MakeUpgrades)
+						manager.RevokeUpgrade(self, up, this);
+				});
+			}
+			else if (building == null)
+			{
+				foreach (var up in info.MakeUpgrades)
+					manager.GrantUpgrade(self, up, this);
+
+				wsb.PlayCustomAnimation(self, info.Sequence, () =>
+				{
+					foreach (var up in info.MakeUpgrades)
+						manager.RevokeUpgrade(self, up, this);
```

Comment:
Since these two lines are written multiple times could you extract them out into appropriately named methods?